orttraining/orttraining/core/framework/ortmodule_graph_builder.cc

```diff
@@ -78,15 +78,16 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
     SetConcreteInputShapes(*input_shapes_ptr);
   }
 
-  // Optimize the inference graph, and if needed, build the gradient graph.
-  std::unordered_set<std::string> x_node_arg_names;
-  ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
+  // This graph will be used only for inferencing, so stop right here.
+  // No need to apply the optimizations for training or build a gradient graph.
   if (!config_.build_gradient_graph) {
-    // This graph will be used only for inferencing, so stop right here.
-    // No need to build a gradient graph.
-    return Status::OK();
+    // Save a copy of inference optimized model
+    return Model::Load(gradient_model_->ToProto(), inference_optimized_model_, nullptr, *logger_);
   }
 
+  // Optimize the inference graph and build the gradient graph.
+  std::unordered_set<std::string> x_node_arg_names;
+  ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
```

Comment:
Out of curiosity, why optimize inference graph for a training scenario? Is it expected an inference graph to be used on a training graph? Would that trigger the issue it is being fixed on the training-only scenario? 

AFAIK, ORTModule builds 2 independent graphs during first forward call, one for training and for inference, so having both graphs in the gradient builder is odd

If that is the case, the unit test below does not cover `ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));` for training-only and inference-only. It does not cover the mixed mode because two independent graphs are used

Reply:
The OptimizeInferenceGraph actually applies training optimizations. Not sure who name that function as such name. We never need to apply inference optimization by our own. It happens during InferenceSession initialization right before the execution.

Reply:
I ran the model one step for training, one step for eval, and get the models for both training and eval for checking. And I can see the graph is different, one has BNInternal and gradient graph, and the other has BN and no gradient graph. Why did you say it doesn't work?

Reply:
> The OptimizeInferenceGraph actually applies training optimizations. Not sure who name that function as such name. We never need to apply inference optimization by our own. It happens during InferenceSession initialization right before the execution.

😃 This was done by me sometime last year when I was splitting the ORTModule code to work with training as well as eval mode. Sorry for the bad naming.